service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -925,6 +973,82 @@ protected void onCommandConnectionEstablished(final HonoConnection commandConnec
+     * @param authenticatedDevice The device that has authenticated to this protocol adapter.
+     *            <p>
+     *            If not {@code null} then the authenticated device is compared to the given tenant and device ID. If
+     *            they differ in the device identifier, then the authenticated device is considered to be a gateway
+     *            acting on behalf of the device.
+     * @param context The currently active OpenTracing span that is used to trace the operation.
+     * @return The assertion.
+     * @throws NullPointerException if any of tenant or device ID are {@code null}.
+     */
+    protected final Future<JsonObject> updateLastGateway(
+            final JsonObject assertRegistrationJson,
+            final String tenantId,
+            final String deviceId,
+            final Device authenticatedDevice,
+            final SpanContext context) {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
+
+        if (!isGatewaySupportedForDevice(assertRegistrationJson)) {
+            return Future.succeededFuture(assertRegistrationJson);
+        }
+
+        final Future<String> gatewayId = getGatewayId(tenantId, deviceId, authenticatedDevice);
+
+        return gatewayId
+                .compose(gwId -> getDeviceConnectionClient(tenantId))
+                .compose(client -> client.setLastKnownGatewayForDevice(deviceId,
+                        Optional.ofNullable(gatewayId.result()).orElse(deviceId), context))
+                .map(r -> assertRegistrationJson);
+    }
+
+    /**
+     * Gets an assertion for a device's registration status and also updates the last gateway associated with the
+     * given device.
+     * <p>
+     * The returned JSON object may include <em>default</em>
+     * values for properties to set on messages published by the device under
+     * property {@link RegistrationConstants#FIELD_PAYLOAD_DEFAULTS}.
+     *
+     * @param tenantId The tenant that the device belongs to.
+     * @param deviceId The device to get the assertion for.
+     * @param authenticatedDevice The device that has authenticated to this protocol adapter.
+     *            <p>
+     *            If not {@code null} then the authenticated device is compared to the given tenant and device ID. If
+     *            they differ in the device identifier, then the authenticated device is considered to be a gateway
+     *            acting on behalf of the device.
+     * @param context The currently active OpenTracing span that is used to
+     *                trace the retrieval of the assertion.
+     * @return The assertion.
+     * @throws NullPointerException if any of tenant or device ID are {@code null}.
+     * @see #getRegistrationAssertion(String, String, Device, SpanContext)
+     * @see #updateLastGateway(JsonObject, String, String, Device, SpanContext)
+     */
+    protected final Future<JsonObject> getRegistrationAssertionWithLastGatewayUpdate(
```

Comment:
why do we need this extra method? Can't we simply use/alter the existing *getRegistrationAssertion* method?

Reply:
Well, yes. We could use the existing `getRegistrationAssertion` method.
My idea here was to make it more explicit what is actually being done in this method. 
But, on second thought, seeing that the long name is somewhat unwieldy and that M5 migration for existing custom protocol adapters will be easier using the existing method, I can change that, adding a note in the javadoc about the extra functionality.